core/router/gateway/monitor.go

```diff
@@ -1,102 +1,191 @@
-	ctx = context.WithValue(ctx, "token", g.Token)
-	return ctx
+	return metadata.NewContext(context.Background(), metadata.Pairs(
+		"id", g.ID,
+		"token", g.Token,
+	))
 }
 
 type monitorConn struct {
-	client pb_noc.MonitorClient
+	clients map[string]pb_noc.MonitorClient
 
-	uplink struct {
-		client pb_noc.Monitor_GatewayUplinkClient
-		sync.Mutex
-	}
-	downlink struct {
-		client pb_noc.Monitor_GatewayDownlinkClient
-		sync.Mutex
-	}
 	status struct {
-		client pb_noc.Monitor_GatewayStatusClient
-		sync.Mutex
+		streams map[string]pb_noc.Monitor_GatewayStatusClient
+		sync.RWMutex
 	}
-}
-
-func (g *Gateway) SetMonitor(client pb_noc.MonitorClient) {
-	g.monitor = &monitorConn{client: client}
-}
 
-func (g *Gateway) uplinkMonitor() (client pb_noc.Monitor_GatewayUplinkClient, err error) {
-	g.monitor.uplink.Lock()
-	defer g.monitor.uplink.Unlock()
-
-	if g.monitor.uplink.client != nil {
-		return g.monitor.uplink.client, nil
+	uplink struct {
+		streams map[string]pb_noc.Monitor_GatewayUplinkClient
+		sync.RWMutex
 	}
-	return g.connectUplinkMonitor()
-}
-
-func (g *Gateway) downlinkMonitor() (client pb_noc.Monitor_GatewayDownlinkClient, err error) {
-	g.monitor.downlink.Lock()
-	defer g.monitor.downlink.Unlock()
 
-	if g.monitor.downlink.client != nil {
-		return g.monitor.downlink.client, nil
+	downlink struct {
+		streams map[string]pb_noc.Monitor_GatewayDownlinkClient
+		sync.RWMutex
 	}
-	return g.connectDownlinkMonitor()
 }
 
-func (g *Gateway) statusMonitor() (client pb_noc.Monitor_GatewayStatusClient, err error) {
-	g.monitor.status.Lock()
-	defer g.monitor.status.Unlock()
-
-	if g.monitor.status.client != nil {
-		return g.monitor.status.client, nil
-	}
-	return g.connectStatusMonitor()
+func (g *Gateway) SetMonitors(clients map[string]pb_noc.MonitorClient) {
+	g.monitor = NewMonitorConn(clients)
 }
 
-func (g *Gateway) connectUplinkMonitor() (client pb_noc.Monitor_GatewayUplinkClient, err error) {
-	if client, err = connectUplinkMonitor(g.monitorContext(), g.monitor.client); err != nil {
-		return nil, err
-	}
-	g.monitor.uplink.client = client
-	return client, nil
+func NewMonitorConn(clients map[string]pb_noc.MonitorClient) (conn *monitorConn) {
+	conn = &monitorConn{clients: clients}
+	conn.uplink.streams = map[string]pb_noc.Monitor_GatewayUplinkClient{}
+	conn.downlink.streams = map[string]pb_noc.Monitor_GatewayDownlinkClient{}
+	conn.status.streams = map[string]pb_noc.Monitor_GatewayStatusClient{}
+	return conn
 }
 
-func (g *Gateway) connectDownlinkMonitor() (client pb_noc.Monitor_GatewayDownlinkClient, err error) {
-	if client, err = connectDownlinkMonitor(g.monitorContext(), g.monitor.client); err != nil {
-		return nil, err
+func (g *Gateway) pushStatusToMonitor(ctx log.Interface, name string, status *pb.Status) (err error) {
+	defer func() {
+		switch err {
+		case nil:
+			ctx.Info("Pushed status to monitor")
+
+		case io.EOF:
+			ctx.Info("Stream closed, retrying")
```

Comment:
There is only one stream to be closed and the function is called with the `ctx`, which contains the monitor name
what else should I add?